File: apps/nursery-web/src/lib/notification-deep-link.ts
/**
 * QA-22 — notification deep-link helpers (Customer Web).
 * Prefer server `data.route` when it is a safe relative path.
 */

export function notificationHref(data: unknown, audience: "customer" | "admin" = "customer"): string | null {
  if (!data || typeof data !== "object") return null;
  const d = data as Record<string, unknown>;
  const route = d.route;
  if (typeof route === "string" && /^\/[A-Za-z0-9/_-]*$/.test(route)) {
    // QA-36-005: customer app has /account/returns/*, not bare /returns/*.
    if (audience === "customer" && /^\/returns(\/|$)/.test(route)) {
      return route.replace(/^\/returns/, "/account/returns");
    }
    return route;
  }
  if (d.order_id != null && String(d.order_id).match(/^\d+$/)) {
    return audience === "admin" ? `/orders/${d.order_id}` : `/account/orders/${d.order_id}`;
  }
  if (d.return_id != null && String(d.return_id).match(/^\d+$/)) {
    return audience === "admin" ? `/returns/${d.return_id}` : `/account/returns/${d.return_id}`;
  }
  if (typeof d.product_slug === "string" && /^[a-z0-9-]+$/i.test(d.product_slug)) {
    return `/product/${d.product_slug}`;
  }
  if (typeof d.campaign_slug === "string" && /^[a-z0-9-]+$/i.test(d.campaign_slug)) {
    return `/campaigns/${d.campaign_slug}`;
  }
  return null;
}

File: apps/nursery_app/lib/services/notification_deep_link.dart
/// QA-22 notification deep links (Customer Mobile).
library;

String? notificationDeepLink(Map<String, dynamic>? data) {
  if (data == null) return null;
  final route = data['route']?.toString();
  if (route != null && RegExp(r'^/[A-Za-z0-9/_-]*$').hasMatch(route)) {
    // Normalize customer web-style / admin-style routes to mobile routes.
    if (route.startsWith('/account/orders/')) {
      return route.replaceFirst('/account/orders/', '/orders/');
    }
    // QA-36-005: bare /returns/* → customer returns UI.
    if (route.startsWith('/returns')) {
      return route.replaceFirst('/returns', '/account/returns');
    }
    if (route.startsWith('/account/returns/')) {
      return route;
    }
    return route;
  }
  final orderId = data['order_id'];
  if (orderId != null) return '/orders/$orderId';
  final returnId = data['return_id'];
  if (returnId != null) return '/account/returns/$returnId';
  final productSlug = data['product_slug']?.toString();
  if (productSlug != null && productSlug.isNotEmpty) return '/product/$productSlug';
  final campaign = (data['campaign_slug'] ?? data['slug'])?.toString();
  if (campaign != null && campaign.isNotEmpty) return '/campaigns/$campaign';
  return null;
}

String? adminNotificationDeepLink(Map<String, dynamic>? data) {
  if (data == null) return null;
  final route = data['route']?.toString();
  if (route != null && RegExp(r'^/[A-Za-z0-9/_-]*$').hasMatch(route)) {
    return route;
  }
  final orderId = data['order_id'];
  if (orderId != null) return '/orders/$orderId';
  final returnId = data['return_id'];
  if (returnId != null) return '/returns/$returnId';
  return null;
}
